plugin/healthchecker/filter.go

```diff
@@ -0,0 +1,127 @@
+package healthchecker
+
+import (
+	"sync"
+	"time"
+
+	"github.com/hashicorp/golang-lru"
+	"github.com/miekg/dns"
+)
+
+type (
+	HealthCheckFilter struct {
+		cache    *lru.Cache
+		checker  Checker
+		interval time.Duration
+		names    map[string]struct{}
+	}
+
+	entry struct {
+		dnsRecord dns.RR
+		healthy   bool
+		quit      chan struct{}
+	}
+)
+
+var mtx sync.Mutex
+
+func NewHealthCheckFilter(checker Checker, size int, interval time.Duration, names map[string]struct{}) (*HealthCheckFilter, error) {
+	cache, err := lru.NewWithEvict(size, func(key interface{}, value interface{}) {
+		mtx.Lock()
+		if e, ok := value.(entry); ok {
```

Comment:
I'm not sure but still since we store `entry` as struct (not pointer) the following situation may happen:
1. Start evicting (have key and value (`entry` struct)) (it already [removed from the cache](https://github.com/hashicorp/golang-lru/blob/master/simplelru/lru.go#L170-L177)) 
2. Tick happens (lock on updating, create new `entry` struct and add it to cache)
3. Continue evicting (close the channel)
4. Have records with closed channel in the cache (potential panic)

Reply:
In this case at step (2) `p.cache.Contains(rec.String())` will return `false` (as you mentioned, evict function invoked after removal) and new `entry` won't be added to the cache, isn't it?

Reply:
Whoops. Yes, you are right

Reply:
I think Denis means we are trying to `close(quit)` after removal of the value from the cache.

Reply:
Yep, if I think correctly, the shouldn't be re-added after starting of evicting. 
But there is another moment: we are trying to `close(quit)` after removal of entry with this channel